instrumentation/google.golang.org/grpc/otelgrpc/test/grpc_stats_handler_test.go

```diff
@@ -1316,3 +1317,27 @@ func checkServerMetrics(t *testing.T, reader metric.Reader) {
 
 	metricdatatest.AssertEqual(t, expectedScopeMetric, rm.ScopeMetrics[0], metricdatatest.IgnoreTimestamp(), metricdatatest.IgnoreValue())
 }
+
+func TestStatsHandlerConcurrentSafe(t *testing.T) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err, "failed to open port")
+	client := newGrpcTest(t, listener,
+		[]grpc.DialOption{
+			grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
+		},
+		[]grpc.ServerOption{
+			grpc.StatsHandler(otelgrpc.NewServerHandler()),
+		},
+	)
+
+	wg := &sync.WaitGroup{}
+	const n = 100
+	for i := 0; i < n; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			doCalls(client)
```

Comment:
This is not reproducing the race. 

I guess that the race could happen when `SendMsg` and `RecvMsg` are executed in separate goroutines.

I think you should make some custom calls on `client` instead of using `doCalls(client)`.

Reply:
Hmm so I am actually struggling to reproduce the data race. I've tried both of these:
```
func TestStatsHandlerConcurrentSafe(t *testing.T) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	require.NoError(t, err, "failed to open port")
	client := newGrpcTest(t, listener,
		[]grpc.DialOption{
			grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
		},
		[]grpc.ServerOption{
			grpc.StatsHandler(otelgrpc.NewServerHandler()),
		},
	)

	ctx := context.Background()

	wg := &sync.WaitGroup{}
	stream, err := client.FullDuplexCall(ctx)
	require.NoError(t, err)
	const n = 100
	wg.Add(2)
	go func() {
		defer wg.Done()
		for i := 0; i < n; i++ {
			const reqSize = 10
			pl := interop.ClientNewPayload(testpb.PayloadType_COMPRESSABLE, reqSize)
			respParam := []*testpb.ResponseParameters{
				{
					Size: reqSize,
				},
			}
			req := &testpb.StreamingOutputCallRequest{
				ResponseType:       testpb.PayloadType_COMPRESSABLE,
				ResponseParameters: respParam,
				Payload:            pl,
			}
			require.NoError(t, stream.Send(req))
		}
	}()
	go func() {
		defer wg.Done()
		for i := 0; i < n; i++ {
			_, err := stream.Recv()
			require.NoError(t, err)
		}
	}()
	wg.Wait()
	require.NoError(t, stream.CloseSend())
	_, err = stream.Recv()
	require.ErrorIs(t, err, io.EOF)
}
```

and

```
func TestStatsHandlerConcurrentSafe(t *testing.T) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	require.NoError(t, err, "failed to open port")
	client := newGrpcTest(t, listener,
		[]grpc.DialOption{
			grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
		},
		[]grpc.ServerOption{
			grpc.StatsHandler(otelgrpc.NewServerHandler()),
		},
	)

	ctx := context.Background()

	wg := &sync.WaitGroup{}
	stream, err := client.FullDuplexCall(ctx)
	require.NoError(t, err)
	const n = 100
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			wg := &sync.WaitGroup{}
			wg.Add(2)
			go func() {
				defer wg.Done()
				const reqSize = 10
				pl := interop.ClientNewPayload(testpb.PayloadType_COMPRESSABLE, reqSize)
				respParam := []*testpb.ResponseParameters{
					{
						Size: reqSize,
					},
				}
				req := &testpb.StreamingOutputCallRequest{
					ResponseType:       testpb.PayloadType_COMPRESSABLE,
					ResponseParameters: respParam,
					Payload:            pl,
				}
				require.NoError(t, stream.Send(req))
			}()
			go func() {
				defer wg.Done()
				_, err := stream.Recv()
				require.NoError(t, err)
			}()
			wg.Wait()
			require.NoError(t, stream.CloseSend())
			_, err = stream.Recv()
			require.ErrorIs(t, err, io.EOF)
		}()
	}
	wg.Wait()
}
```
Would appreciate it if you have any pointers.

Reply:
the other thing i was trying is to try to simulate a race between CloseSend and Recv, which is more likely what's happening given the location of the code. that still doesn't repro:
```
func TestStatsHandlerConcurrentSafe(t *testing.T) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	require.NoError(t, err, "failed to open port")
	client := newGrpcTest(t, listener,
		[]grpc.DialOption{
			grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
		},
		[]grpc.ServerOption{
			grpc.StatsHandler(otelgrpc.NewServerHandler()),
		},
	)

	ctx := context.Background()

	wg := &sync.WaitGroup{}
	const n = 100
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			wg := &sync.WaitGroup{}
			wg.Add(2)
			stream, err := client.FullDuplexCall(ctx)
			require.NoError(t, err)
			go func() {
				defer wg.Done()
				sendWg := &sync.WaitGroup{}
				for i := 0; i < n; i++ {
					sendWg.Add(1)
					go func() {
						defer sendWg.Done()
						const reqSize = 10
						pl := interop.ClientNewPayload(testpb.PayloadType_COMPRESSABLE, reqSize)
						respParam := []*testpb.ResponseParameters{
							{
								Size: reqSize,
							},
						}
						req := &testpb.StreamingOutputCallRequest{
							ResponseType:       testpb.PayloadType_COMPRESSABLE,
							ResponseParameters: respParam,
							Payload:            pl,
						}
						require.NoError(t, stream.Send(req))
					}()
				}
				sendWg.Wait()
				require.NoError(t, stream.CloseSend())
			}()
			go func() {
				defer wg.Done()
				for i := 0; i < n; i++ {
					_, err := stream.Recv()
					require.NoError(t, err)
				}
				_, err = stream.Recv()
				require.ErrorIs(t, err, io.EOF)
			}()
			wg.Wait()
		}()
	}
	wg.Wait()
}
```

Reply:
nevermind, I reproed it - it only happens if client cancels context in the middle of streaming